src/NHibernate/Id/Enhanced/OptimizerFactory.cs

```diff
@@ -365,26 +395,37 @@ public PooledLoOptimizer(System.Type returnClass, int incrementSize) : base(retu
 				}
 			}
 
+			public class GenerationState
+			{
+				public long LastSourceValue { get; internal set; } = -1;
+				public long Value { get; internal set; }
+			}
+
 			public override object Generate(IAccessCallback callback)
 			{
 				using (_asyncLock.Lock())
 				{
-					if (_lastSourceValue < 0 || _value >= (_lastSourceValue + IncrementSize))
+					var generationState = _stateStore.LocateGenerationState(callback.GetTenantIdentifier());
+
+					if (generationState.LastSourceValue < 0 || generationState.Value >= (generationState.LastSourceValue + IncrementSize))
 					{
-						_lastSourceValue = callback.GetNextValue();
-						_value = _lastSourceValue;
+						generationState.LastSourceValue = callback.GetNextValue();
+						generationState.Value = generationState.LastSourceValue;
 						// handle cases where initial-value is less than one (hsqldb for instance).
-						while (_value < 1)
-							_value++;
+						if (generationState.Value < 1)
+						{
+							generationState.Value = 1;
+						}
 					}
 
-					return Make(_value++);
+					return Make(generationState.Value++);
 				}
 			}
 
+
 			public override long LastSourceValue
 			{
-				get { return _lastSourceValue; }
+				get { return _stateStore.NoTenantGenerationState.LastSourceValue; }
```

Comment:
It should return -1 if not initialized. See https://github.com/nhibernate/nhibernate-core/blob/79ad5b8f28490dc0818d9329bb079451b0a9800a/src/NHibernate/Id/Enhanced/IOptimizer.cs#L24-L26

Your code throws. New method also doesn't return -1. It generates new state.